README.md
# 写作天下

## 简介

为作家们创造世界而诞生，执云作笔，诉尽平生意。

集简约UI与人性化AI于一体的码字工具，无论是小说、作文、日记、报告，都能轻松驾驭。

**QQ交流群：705849222**



---



## 功能

1. **自由的目录**：自动序号，拖拽排序，随意调整
2. **自定义界面**：全局自定义主题、风格，可使用CSS代码
3. **智能标点系统**：独有的强大又省力的标点 AI
   1. **智能引号**：增删移动全在一键之间
   2. **智能空格**：缩进、标点、移动光标融为一体
   3. **智能回车**：双引号内回车、自动添加句末标点、自动缩进
   4. **自动标点**：语气词后面自动加上标点符号
4. **名片库系统**：为文中所有人物、事物创建名片
   1. **名字高亮**，直观高效
   2. **鼠标悬浮**：随时查看简介与详情
   3. **章节出现**的名片一览
   4. **快速搜索**所有名片
5. **云端同步**：稳定、全自动，多一重保险。支持同步到手机。
6. **云排行榜**：所有用户在线拼字
7. **多线大纲**：主线/辅线/事件线/感情线…等
8. **语言高亮**：人物语言描写高亮显示
9. **章内标注**：章节内埋坑备注，可作细纲，导出、发布时自动去除
10. **高潮模式**：自动识别当前情节激烈程度，微调小说AI
11. **导入导出**：花样自定义格式，便捷的多文件拖拽导入导出
12. **字数统计**：精确到几个空格的数量
13. **同音词覆盖**：选词错误，无需删除，直接覆盖旧内容
14. **敏感词检测**：高亮所有敏感词，可自定义词库
15. **多功能替换**：简繁转换、敏感词替换、敏感词转拼音等
16. **自动分段**：换行/排版时，过长段落自动在合适的地方分段
17. **词语提示**：输入时自动提示同义词、相关词、常用句
18. **随机取名**：最轻便的取名方式，要啥来啥
19. **一键排版**：全自动，包括英文空格、首字母大写、长段落自动分段等
20. **资源搜索**：全局浏览所有资源文件，快速搜索
21. **小黑屋**：多种模式强制码字
22. **剪贴板**：保存所有复制的历史记录
23. **回收板**：记录所有删除/剪切的长文本，避免丢失
24. **小说类型**：只需小说名便可判断小说是什么类型
25. **输入动画**：文字平滑输入、光标平滑移动
26. **微动背景**：在不影响注意力的情况下，微调背景颜色
27. **情绪滤镜**：智能分析人物情绪，添加相应背景滤镜
28. **场景图片**：识别故事环境，设置对应的背景图片
29. **错词替换**：错别字高亮和一键替换
30. **汉语词典**：内置完整词典，显示所有释义，附加成语接龙功能
31. **主题管理**：强大的主题功能，切换色彩、词库、CSS样式、动态背景




## 以用户为本的小细节

《写作天下》是一款真正以用户为本的应用，功能强大的同时上手容易，几乎无难度。

以下这些或许难以察觉的小设计更是打造了极度良好的用户体验：

1. **创建新书名字**：分析用户之前的各种操作，获取可能取名的名字
2. **自动章节序号**：中间临时插入章节、随意排序，而无需修改后面全部序号
3. **目录整十对齐**：使目录中文序号的排列整齐，避免长度差导致的错落
4. **还原光标位置**：调整窗口大小、修改字体大小后，光标仍在窗口同比例位置
5. **排版输入位置**：输入聚焦在原来的位置不变，甚至连滚动进度也一样
6. **恢复编辑位置**：重启程序，滚动位置、光标位置、甚至连选中的内容都恢复原态
7. **独特动态界面**：可视控件全部重写，哪怕是一个小小的按钮都如此与众不同
8. **智能引号细节**：第二句人物描写的双引号前缀为逗号，而非冒号
9. **动态波浪动画**：没有打开作品时为不显得空洞，播放动画
10. **趣味交互按钮**：各种线条律动、缩放特效，活灵活现
11. **小说类型判断**：输入书名，自动分析作品类型，无需手选~
12. **多段落后引号**：智能回车仅最后一段有后引号，语法准确
13. **小黑屋可逃离**：当检测到用户无法完成小黑屋时，提供逃离方式（临时有效）
14. **点不到的按钮**：会跑的“强制退出”按钮！点到算你手快
15. **长段落自动分段**：不拆分语言描写，并且尽量每段字数接近
16. **小黑屋字数预测**：根据使用习惯，猜测几个可能的数字
17. **自动打字机模式**：在末尾码字时自动开启打字机模式
18. 针对高端用户，**自定义所有**词库、界面CSS代码、菜单格式、配置INI文件等
19. 窗口一体化，可隐藏不需要的控件，通过搜索面板直接**输入指令并执行**
20. 其他彩蛋，比如夜间码字温馨鼓励语、成语接龙



## 快捷键

- Ctrl + ←/→    按词语移动
- Alt + ←/→    按句子移动
- Ctrl + Alt + ←/→    按段落移动
- Ctrl + Shift + ↑/↓    扩大/缩小选择
- Ctrl/Alt + ↑/↓    屏幕滚动，alt更快
- 空格键    智能空格
- 引号键    智能引号
- 回车键    智能回车
- Tab键    句子补全+光标跳过
- Ctrl+T    一键排版
- Ctrl+D    句内同音词替换（不必连续）
- Ctrl+F    章内文字搜索
- Ctrl+P    全局综合搜索
- Ctrl+B    剪贴板
- Ctrl+J    文字跳转

> 支持自定义快捷键
>



![截图0：微动背景，时空渐变](resources/主题预览/主题：默认.png)

![截图1](resources/主题预览/主题：爱丽丝蓝.jpg)

![截图2](resources/主题预览/主题：薄荷清新绿.jpg)

![截图3](resources/主题预览/主题：雾玫瑰红.jpg)

![截图4](resources/主题预览/主题：薰衣草紫.jpg)

![截图5](resources/主题预览/主题：玉米丝黄.jpg)

![截图6](resources/主题预览/主题：白银灰.jpg)



---

# 部署

1. 安装 `Qt5.13.1` （理论上相近版本亦可）
2. 使用 `Qt Creator` 打开 `WriterFly.pro` 文件
3. 手动拷贝资源文件"安装包/data"到运行目录（可选）
4. 运行



# 技术特点

开发环境：C++/Qt5.13.1，Qt Creator，Windows / MacOS / Linux

运行环境：Windows、MacOS、Linux、Android、iOS




## 1、标点AI

虽然是简单暴力的枚举（都是自己遇到的），但是有特判了**上千种情况**，覆盖日常绝大部分

若是人物语言描写，还能根据**人物的表情动作神态**来分析情感程度，进一步提高准确度。

手动判断，故覆盖不全，不过在一定情况下，枚举简单易用，是为上上之选！

配合分析情节激烈程度的“**高潮模式**”，使智能标点更加准确。

还有各种常用按键的自动化操作，极大程度上增加效率！



## 2、仿代码编辑器

当今所谓的小型写作软件，单论写作方面，其实和一个记事本没多少差别，无非多了目录与自动缩进，以及简单的自动保存等。其一些亮点，比如强制写作的小黑屋、多人竞争的在线拼字，都无法从“码字”本身来帮助作者达到更高效的办公，这些附加的功能甚至会让作者分心，无法好好**静心创作**。

写作天下定位一款单纯的“**文学创作**”编辑器，主打**从“写作”本身提高效率，弱化与文字无关的功能**。或许后期将会添加小黑屋、拼字等，但绝不会将这些作为主要功能。

> 备注：近期繁忙，无力开发，所以只是“思路”！

一些功能：

- 增强的**编辑功能**

- **自动提示**
- **面向对象**写作：名片系统
- 文字高亮（名片高亮）
- 高度个性化设置（功能细节、快捷键、主题等）

### 追求自动

一句话：**能自动的，绝不手动！**

为了实现更加人性化的功能，程序中使用了大量的算法。

比如**括号匹配**功能，输入左括号时自动添加右括号、删除键删除成对的符号，这里使用数据结构中的**栈**，判断光标前后各自左右括号的数量，入栈出栈，来判断是否需要添加/删除，而不仅仅依靠光标前一个字符。

还有光标移动、在标点前面换行、在语言描写内回车自动插入前后引号、修改引号前面的逗号与冒号等，**一个按键，多种功能**。

### 轻松提示

经过多种方案的性能测试，选取了速度最快、运行最稳定的方式。



## 3、平滑输入

创作，或是为了宁静的内心，或是为了美好的生活，终是需要解放内心的压抑。

而写作天下的输入动画与平滑光标，为了打造**愉悦的创作环境**，可谓是呕心沥血。

![平滑输入(图中还有自动提示)](resources/截图/平滑输入.gif)



## 4、微动背景+情绪滤镜+场景图片

一直更换背景？何不试试新时代的纯色背景！

可能不是一套实用的功能，但是真的很好玩！

- **微动背景**：在用户体验不到的情况下，在一定区域内缓慢变动细节，同一个窗口，不一样的色彩，百看不厌
- **情绪滤镜**：输入不同的句子，实时分析其中的情绪（如愉悦、哀伤、寂寞）和色彩（如红色、葡萄、奶油），并为背景套上一层色彩蒙版
- **场景图片**：同上，分析不同的事物，如苍穹、江河、松柏、晚霞、夜、冰霜、雷雨、墓地、战争等（目前自带10类，可扩展）

这些背景均**可重叠**，JPG+背景色+PNG+情绪滤镜+场景图片，五种效果可同时使用！

最重要的是，**可扩展性**，通过主题来增加词库和背景，例如最新的国庆主题，除了红红火火的背景色，就有个国庆的天安门图片，输入“国庆”、“十一”等关键词唤出



## 5、其他

学习用户习惯的小黑屋数字预测、基于TF-IDF文本相似度的正文-故事线互相定位、活灵活现的交互按钮库等

CHANGELOG.md
# 写作天下 各版本更新内容


### 2019.12.20  -  1052(v1.5.5)

更新方向：名片AI

- 快速设置名片标签的按钮
- 创建名片时根据已有名片自动设置颜色
- 创建名片时自动设置最合适的标签
- 自动为不同标签设置最合适的颜色
- 修复无法选择作品封面的问题



### 2019.12.8  -  1051(v1.5.4)

更新方向：设置界面、主题切换

- 完全重构设置界面，增项更多设置项
- 添加设置分组切换动画
- 设置分割线改为“不常用项”
- 新增6套内置主题
- 优化关闭当前行背景不需要修改内容
- 优化切换主题时读取点击水波纹动画
- 修复未保存CSS样式表到主题
- 修复设置界面背景大变的问题
- 修复名片导致的程序崩溃问题



### 2019.12.1  - 1050(v1.5.3)

- 新增可视区域光标跳转
- 优化大量夜间模式白色控件
- 优化收起侧边栏时聚焦到正文输入框
- 修复背景图模糊无法取消的问题
- 微调标题栏模糊，新增设置项



### 2019.11.23  - 1049(v1.5.2)

更新方向：名片优化

- 新增正文菜单中修改名片颜色
- 新增保存主题时填充默认名字
- 新增名片自动滚动至简介开关
- 新增保存名片框位置和大小
- 新增名片框跟随主窗口移动
- 新增Ctrl键移动多个名片对话框
- 新增shift键移动子名片对话框
- 修复标题栏模糊无效的问题
- 修复正文修改名片颜色无法保存
- 修复修改主题系列颜色多个错误
- 修复delete键无效
- 修复在作品相关插入章节序号问题
- 调整名片框背景颜色为副色调
- 美化菜单选中情况
- 优化搜索结果滚动条
- 优化光标附近没有名片时禁用名片菜单



### 2019.11.17  -  1048(v1.5.1)

更新方向：窗口界面设计

- 新增全局窗口圆角效果
- 新增窗口边界阴影效果
- 新增恢复窗口边缘拖动
- 新增按标签、名字、信息搜索名片
- 新增名片搜索结果美化
- 新增所有对话框可调整大小
- 新增添加名片右键菜单在没有名片库时禁用
- 修复没有名片库时编辑器右键菜单崩溃的问题
- 修复名片界面移动过快消失的问题
- 修复主题界面风格问题
- 默认关闭标题栏模糊
- 部分界面细节优化



### 2019.11.10  -  1047(v1.5.0)

更新方向：动画、名片库

- 新增名片库系统:
  - 侧边栏名片控件
  - 名片文字高亮
  - 名片简介与详情
  - 鼠标悬浮显示简介
  - Ctrl键下划线
  - 菜单快速操作
  - 名片搜索结果
- 新增启动摇摆动画
- 新增背景图片模糊
- 新增非线性点击动画
- 优化一部分窗口动画
- 优化字数统计的格式
- 优化切换故事线焦点
- 修复故事线无法删除
- 修复打开章节滚动到中间
- 默认关闭鼠标点击水波纹



### 2019.10.26  -  1046(v1.4.3)

更新方向：交互通知、快捷键

- 添加内测渠道版本
- 新增通知控件
- 新增更新通知询问
- 新增快捷键列表
- 新增自定义快捷键
- 新增小黑屋字数询问
- 新增拖拽全书txt直接导入
- 新增离线升级功能
- 部分控件细节优化



### 2019.10.1  -  1045(v1.4.2)

更新方向：小黑屋信息

- 新增小黑屋信息显示窗口
- 新增或许能点到的退出小黑屋按钮
- 优化一部分标点词库



### 2019.9.30  -  1044(v1.4.1)

更新方向：主题

- 新增支持主题管理系统
- 新增支持自定义图标
- 新增支持自定义情绪滤镜
- 新增支持自定义场景图片
- 新增支持自定义按钮CSS样式
- 优化多显示器启动和最大化窗口
- 优化启动时缩小过大的窗口尺寸
- 优化目录点击空白处的操作和菜单
- 优化背景颜色透明时各项控件背景色
- 优化切换背景图片时自动调整背景透明度
- 修复背景图片导致标题栏闪屏的问题
- 修复背景透明导致的一系列透明问题
- 修复定位到故事线选择区域消失问题
- 修复后台云同步导致更新失败的问题
- 修复没有章节时综合搜索崩溃的问题



### 2019.9.24  -  1043(v1.4)

注意：升级后需重新登录同步云端积分，本地积分将清零一次

- 新增修改账号、昵称、密码
- 新增存储数据到通用路径
- 新增从章节自动定位到故事线
- 新增根据使用习惯预测小黑屋
- 新增敏感词替换为拼音
- 新增敏感词替换为首字母
- 添加开始菜单快捷方式
- 放松字数限制，昵称最少两位
- 优化键盘快捷键平滑滚动
- 优化定位到目录展开侧边栏
- 修复退出时长期后台运行的问题
- 修复对话框两次背景动画的问题
- 修复部分按钮点击后背景灰色的问题



### 2019.9.13   -  1042(v1.3.2)

更新方向：小说数据

- 新增登录时检测排行榜
- 新增退出时后台云同步
- 新增定时同步其他设备
- 新增导入TXT指定编码
- 新增记录每日字数和时间
- 新增编辑器水波纹移动消散
- 优化导入章节名默认表达式
- 优化导入选择分卷逻辑判断
- 优化导出自由选择换行方式
- 优化自定义菜单的CSS样式
- 优化一小部分常用标点词库
- 修复注册按钮文字错误
- 修复注册后未开始上传问题
- 修复旧导出文件被覆盖问题
- 修复目录无法拖拽排序问题
- 修复tab控件切换背景问题
- 修复云端删除作品失败卡住
- 修复侧边栏不透明背景问题
- 修复微动背景编程白色问题
- 修复夜间模式登录背景问题



### 2019.9.5  -  1041(v1.3.1)

更新方向：交互操作

- 新增小黑屋开机恢复
- 新增切换为菜单栏模式
- 新增搜索指令参数模式
- 新增完全自定义菜单脚本
- 新增标题栏模糊自动开关
- 新增综合搜索可指定分类
- 新增连续的快速成语接龙
- 新增小黑屋逃离（联系开发者）
- 优化调整窗口大小时固定光标
- 优化修改字体大小时固定光标
- 优化对话框Tab键聚焦顺序
- 修复隐藏窗口按钮无法移动
- 修复关闭窗口时界面消失问题
- 修复隐藏侧边栏后标题栏位置
- 修复改变字体大小进度条问题
- 修复退出全屏时无法保存最大化
- 修复章节重命名后存储格式问题



### 2019.8.30  -  1040(v1.3)

更新方向：文本转换

- 新增上传积分
- 新增文本替换
- 新增错别字高亮
- 新增错别字纠错
- 新增敏感词替换
- 新增侧边栏隐藏
- 新增上传码字速度
- 新增搜索汉语词典
- 新增搜索执行指令
- 新增隐藏窗口按钮
- 优化搜索快捷键
- 优化章节搜索结果
- 优化夜间模式颜色
- 优化原生样式Dock
- 优化同步弹窗适应主题
- 默认开启平滑滚动和输入动画
- 修复设置项描述不变色的问题



### 2019.8.26  -  1039(v1.2.2)

更新方向：云同步

- 修复下载码字风云目录为空的错误
- 新增云同步，并非只上传
- 新增下载目录为空时询问，而非停止
- 新增小黑屋，支持字数/时间，警告/强制
- 优化正文搜索上下按钮可用性
- 优化搜索结果居中
- 优化搜索滚动时光标动画
- 修复鼠标右键拖动
- 修复窗口全屏无法缩小的问题
- 日常优化标点词库





### 2019.8.20  -  1038(v1.2.1)

更新方向：动画

- 新增编辑框模糊到标题栏
- 新增编辑框鼠标点击动画
- 新增显示云同步“比对中”状态
- 新增菜单与搜索栏CSS样式
- 重构目录动画，添加效果
- 优化平滑滚动反向控制
- 修复透明背景时控件的背景问题
- 修复侧边栏切换时带动背景问题



### 2019.8.16  -  1037(v1.2)

更新方向：云同步

- 新增账户登录与云同步
- 新增尝试定位到大纲
- 新增后面添加章节
- 新增简繁转换
- 新增敏感词消除
- 优化平滑滚动细节，包括光标移动动画
- 默认关闭输入动画，避免win10输入法兼容性问题
- 微调编辑器菜单
- 修复目录气泡动画闪动的问题
- 修复空章节默认缩进失效的问题
- 修复大纲光标高度变化大小问题
- 修复作品信息文字显示不全的问题



### 2019.8.3  -  1036(v1.1.1)

- 新增移动正文内容到相邻章节
- 新增移动选区内容到相邻章节
- 新增大纲搜索条
- 修复打开上一章/下一章没跳过目录的问题
- 修复禁止自动保存后关闭程序未保存的问题
- 优化夜间模式侧边栏背景为透明
- 优化目录按钮文字颜色
- 统一所有菜单图标颜色



### 2019.7.30  -  1035(v1.1)

更新方向：文字情景识别

- 新增情景滤镜：智能分析情节中人物情绪、所处环境，为背景添加相应滤镜
- 新增场景图片：识别小说中的故事环境，以对应的图片作为背景图片（极其有限）
- 新增高潮模式：自动判断情节激烈程度，修改一部分AI算法
- 新增高潮模式调节微动背景渐变比例
- 新增目录及其章节直接导入
- 新增章节txt直接拖动到目录导入
- 新增Jieba分词模式（需下载数据包）
- 新增一部分可自定义界面的CSS样式
- 新增目录合并连续的分卷
- 新增目录将分割分卷
- 新增章节直接分割一部分到新章
- 新增章节编辑器处跳转到上下章
- 新增编辑器中定位到章节目录
- 添加一个半夜才有的的彩蛋
- 优化目录插入动画
- 优化运行时背景图片加载速度
- 修复Friso分词失效的问题
- 修复启动时不会自动打开第一章



### 2019.7.26  -  1034(v1.0.1)

- 新增回车键强制另起一段(隐藏设置)
- 新增半夜码字温馨祝福语彩蛋
- 新增复制作品
- 优化自动根据控字体大小调整控件大小
- 修复启动时打开上一章的焦点问题
- 修复打开章节光标丢失问题
- 修复大字体时部分控件排版问题



### 2019.7.23  -  1033(v1.0.0)

- 新增搜索近义词
- 新增码字字数统计
- 新增使用时间统计
- 新增全局文字颜色
- 新增夜间模式切换
- 新增版本更新提示
- 新增下划线出现动画
- 新增微动伪静态背景
- 新增主窗口全屏创作
- 新增本次输入速度统计
- 新增目录显示每章字数
- 新增目录悬浮预览开关
- 新增部分CSS样式表支持
- 新增可交互的创意动态按钮
- 新增目录和编辑器平滑滚动
- 新增创建新章自动递增序号
- 新增上下箭头跳到相邻段落
- 新增首尾上下箭头添加新段落
- 新增自动输入作者和作品类型
- 优化搜索面板帮助
- 优化作品一览界面
- 优化小说类型数据库
- 优化智能标点数据库
- 优化侧边栏布局细节
- 优化目录选中项背景
- 优化搜索条背景阴影
- 优化搜索条按钮动画
- 优化默认开启下划线
- 优化水波浪动画速度
- 优化添加分卷的图标
- 优化目录删除后焦点位置
- 优化输入动画为背景反色
- 优化搜索条背景随主题色变化
- 修改部分文件夹名称
- 修复作品列表颜色问题
- 修复自动更新失效的问题
- 修复大纲输入动画跳动的问题
- 修复侧边栏宽度无法调整的问题
- 减少搜索条隐藏后阴影仍在的问题



### 2019.6.15  -  1032(v0.9.1)

更新方向：数据显示

- 新增剪贴板
- 新增回收板
- 新增敏感词高亮
- 新增人物语言高亮
- 新增部分颜色随点缀色变化
- 新增特别特别详细的字数统计
- 新增目录章节和分卷字数统计
- 新增目录章节鼠标悬浮预览
- 新增搜索面板动画
- 新增搜索面板帮助列表
- 提高下划线默认透明度
- 加速波浪动画速度
- 优化目录项背景显示效果
- 优化搜索条颜色为副色调
- 日常优化标点词库
- 修复动画内存溢出问题
- 修复方形模式按钮问题
- 修复搜索面板自动选中的问题
- 修复搜索面板找不到人名的问题



### 2019.6.13  -  1031(v0.9.0.1)

- 优化选中文字后的光标动画问题
- 修复语气词后面总是逗号的问题
- 调整创建第一部作品的默认书名



### 2019.6.12  -  1030(v0.9)

更新方向：搜索与编辑器

- 新增章节内搜索功能（Ctrl+F）
- 新增全局搜索面板（Ctrl+P）
- 新增平滑滚动动画
- 新增人物语言高亮
- 新增章节标注高亮
- 新增编辑器设置组
- 新增文字下划线
- 新增行间距设置
- 新增当前行高亮开关
- 新增导出时自动去除标注
- 新增创建作品时自动命名
- 新增目录拖拽直接导出分卷或者章节
- 新增标点覆盖支持破折号和省略号
- 优化标点词库
- 扩大Ctrl+D同音词覆盖范围
- 编辑当前作品以外章节时在标题栏显示作品名字
- 优化打开章节光标自动聚焦到第一个非空白符处
- 修复快捷键一键复制无效的问题
- 修复打开章节第一次编辑时无法进行AI判断
- 修复打开章节第一次编辑时光标动画问题
- 修复目录很容易被认为拖拽的问题
- 修复撤销到开始状态时光标跳到开头问题



### 2019.6.2  -  1029(v0.8.4)

更新方向：目录动画

- 新增目录点击动画
- 新增目录悬浮动画
- 调整打开章节动画
- 新增章节内搜索栏
- 新增搜索文字高亮
- 修复新章黑色背景



### 2019.5.31  -  1028(0.8.3)

- 新增没有作品时的多重波浪动画
- 优化目录“不使用分卷”重命名时的编辑框位置
- 修复编辑框菜单“排版”项只在空时有效的问题
- 修复设置项隐藏属性点不开的问题



### 2019.5.30  -  1027(v0.8.2)

- 新增根据作品名字判断小说类型
- 新增自定义全局字体
- 新增目录和大纲切换动画
- 新增Ctrl+滚轮修改大纲字体大小
- 新增部分信息动画
- 新增中文章序整十对齐
- 调整智能空格句号变成逗号
- 美化作品一览界面
- 优化全文尾的tab键缩进
- 修复输入动画的字体颜色问题
- 修复无法修改作品封面的问题
- 修复创建默认故事线集时清空当前故事线的问题

globals/defines.h
/**
 * 全局宏定义
 */

#ifndef DEFINES_H
#define DEFINES_H

#include <QString>


// ====标题栏 ====
#define BUTTON_HEIGHT 30 // 按钮高度
#define BUTTON_WIDTH 30  // 按钮宽度
#define TITLE_HEIGHT 30  // 标题栏高度

// ==== 布局 ====
#define CUTTING_SPACE 0  // 切掉的布局控件间宽度

// ==== 目录 Role ====
enum DRole {
    DRole_CHP_NAME = 0,
    DRole_CHP_NUM = 1,
    DRole_CHP_NUM_CN = 2,
    DRole_CHP_DETAIL = 3,
    DRole_CHP_ISROLL = 10,
    DRole_CHP_ISHIDE = 4,
    DRole_CHP_STEMP = 50,
    DRole_CHP_OPENING = 100,
    DRole_ROLL_SUBS = 11,
    DRole_ROLL_COUNT = 12,
    DRole_CHP_FULLNAME = 13,
    DRole_RC_ANIMATING = 14,
    DRole_RC_SEARCH = 15,
    DRole_CR_COUNT = 16,
    DRole_ROLL_NAMES = 17,
    DRole_CHP_ROLLINDEX = 18,
    DRole_CHP_CHPTINDEX = 19,
    DRole_RC_SELECTING = 20,
    DRole_RC_CLICKPOINT,
    DRole_RC_SELECT_PROG100,
    DRole_RC_UNSELECT_PROG100,
    DRole_RC_HOVERPROG,
    DRole_RC_PREV_HOVERPROG,
    DRole_RC_NEXT_HOVERPROG,
    DRole_RC_WORD_COUNT,
};

#define SHOW_ROLE Qt::UserRole+26
#define TITLE_ROLE Qt::UserRole+58
#define SIZE_ROLE Qt::UserRole + 84

#define DIR_DEVIATE 14      // 目录左边的距离
#define DIR_DEVIATE_ADD 20  // 章节比分卷靠后的距离

// ==== 文件 ====
#define NOVELDIRFILENAME "dir.txt"
#define OUTLINE_DIR_FILENAME "outline_dir.txt"


// ==== 编辑 ====
enum EditType {
    EDIT_TYPE_NONE,   // 没有编辑
    EDIT_TYPE_CHAPTER, // 章节
    EDIT_TYPE_OUTLINE, // 大纲
    EDIT_TYPE_FINE_OL, // 细纲
    EDIT_TYPE_BRIEF,   // 简介
    EDIT_TYPE_DETAIL,  // 细节
    EDIT_TYPE_UNKNOW = 9,  // 未知
};


// ==== 方向 ====
enum DirectType {
    DIRECT_NONE,
    DIRECT_TOP,
    DIRECT_LEFT,
    DIRECT_RIGHT,
    DIRECT_BOTTOM
};


// ==== 控件 ====
#define TRANSPARENT_SHOW_KIND_NOVEL_INFO 3


// ==== 引导 ====
enum LINE_GUIDE {
    LINE_GUIDE_NONE,
    LINE_GUIDE_CREATE_BOOK
};


#endif // DEFINES_H
